### ISSUE TITLE
Issue trying with a SageMaker Notebook: "sagemaker-ssh-helper:SSMManager:SSH Helper not yet started? Retrying."

### ISSUE THREAD

**Issue by mlnrt** [aws-samples/sagemaker-ssh-helper#23]
Hello, this looks like a great project but I have been struggling for a day and half to get it working.
I am trying to do the [Local IDE integration with SageMaker Studio over SSH for PyCharm / VSCode](https://github.com/aws-samples/sagemaker-ssh-helper#studio) but with SageMaker Notebook instead of Studio, and from a Windows environment.

First I have to say that, the fact that the instructions are spread out in many places (in the main README, the account setup page, the FAQ, etc.) do not help with clarity. Maybe, separate step-by-step end-to-end instructions for each use case would help.

I first tried to do this in a corporate environment, where SageMaker notebook are in a private VPC with no direct internet breakout in the account but going through a corporate firewall and proxy and with no IAM user but temporary credentials. As I miserably failed, I reverted to first test it in my personal environment with the SageMaker notebook deployed in the default AWS managed environment but I am still failing.

In my local environment, I think I got all the requirements (installing `sagemaker-ssh-helper`, running the `./sm-local-install-force` script from an admin bash), configuring the AWS environment (SSM and IAM policies as mentioned [here](https://github.com/aws-samples/sagemaker-ssh-helper/blob/main/IAM_SSM_Setup.md), copying and running the `SageMaker_SSH_Notebook.ipynb` notebook.
I also looked at the video for the SageMaker studio case, which is quite different, but according to the instructions the only difference is just the notebook we are execution. So I think I got the instructions correct but as it is failing I guess I am missing something...

I can see in SSM fleet manager the `mi-******` node ID of the SageMaker notebook instance but the notebook script does not display it (not sure if that is normal or not). The last logs I have on the jupyter notebook outputs are:
```
j7h3m4c9pf-algo-1-jq0oi | # Running forever as daemon
j7h3m4c9pf-algo-1-jq0oi | amazon-ssm-agent
j7h3m4c9pf-algo-1-jq0oi | Initializing new seelog logger
j7h3m4c9pf-algo-1-jq0oi | New Seelog Logger Creation Complete
j7h3m4c9pf-algo-1-jq0oi | Applying config override from /etc/amazon/ssm/amazon-ssm-agent.json.
j7h3m4c9pf-algo-1-jq0oi | 2023/04/21 12:20:31 Found config file at /etc/amazon/ssm/amazon-ssm-agent.json.
j7h3m4c9pf-algo-1-jq0oi | 2023/04/21 12:20:31 processing appconfig overrides
j7h3m4c9pf-algo-1-jq0oi | 2023/04/21 12:20:31 Found config file at /etc/amazon/ssm/amazon-ssm-agent.json.
j7h3m4c9pf-algo-1-jq0oi | 2023/04/21 12:20:31 processing appconfig overrides
j7h3m4c9pf-algo-1-jq0oi | Applying config override from /etc/amazon/ssm/amazon-ssm-agent.json.
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO Proxy environment variables:
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO http_proxy: 
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO no_proxy: 
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO https_proxy: 
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO Checking if agent identity type OnPrem can be assumed
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO Agent will take identity from OnPrem
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [amazon-ssm-agent] using named pipe channel for IPC
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [amazon-ssm-agent] using named pipe channel for IPC
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [amazon-ssm-agent] using named pipe channel for IPC
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [amazon-ssm-agent] amazon-ssm-agent - v3.2.815.0
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [amazon-ssm-agent] OS: linux, Arch: amd64
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [amazon-ssm-agent] Starting Core Agent
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [CredentialRefresher] credentialRefresher has started
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [CredentialRefresher] Starting credentials refresher loop
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:31 INFO [CredentialRefresher] Next credential rotation will be in 29.997203369883334 minutes
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:32 INFO [amazon-ssm-agent] [LongRunningWorkerContainer] [WorkerProvider] Worker ssm-agent-worker is not running, starting worker process
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:32 INFO [amazon-ssm-agent] [LongRunningWorkerContainer] [WorkerProvider] Worker ssm-agent-worker (pid:611) started
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:20:32 INFO [amazon-ssm-agent] [LongRunningWorkerContainer] Monitor long running worker health every 60 seconds
j7h3m4c9pf-algo-1-jq0oi | 2023-04-21 12:50:31 INFO [CredentialRefresher] Next credential rotation will be in 29.997979336833332 minutes
```

Then if I try to do a `AWS_PROFILE=default sm-local-ssh-notebook connect <<notebook-instance-name>>`, then I get the following:
```
INFO:sagemaker-ssh-helper:SSMManager:Querying SSM instance IDs for SageMaker notebook instance test
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:sagemaker-ssh-helper:SSMManager:SSH Helper not yet started? Retrying. Seconds left: 300
INFO:sagemaker-ssh-helper:SSMManager:SSH Helper not yet started? Retrying. Seconds left: 290
INFO:sagemaker-ssh-helper:SSMManager:SSH Helper not yet started? Retrying. Seconds left: 280
INFO:sagemaker-ssh-helper:SSMManager:SSH Helper not yet started? Retrying. Seconds left: 270
```
But if I am understanding the instructions correctly this is not the SSH Helper on the local machine, correct? This is the SSH Helper on the SageMaker notebook, right?
Did I missed something obvious?

**Comment by ivan-khvostishkov** [comment]
Hi, @mlnrt, thank you for this detailed and very useful feedback! 

When SSM agent starts, it doesn't print the instance ID, so you might want to look at the previous log lines in the notebook where SSH Helper creates the hybrid activation. The log will look like this: 
```
8yre1xga6h-algo-1-rmi0c | sm-init-ssm: Detected SageMaker resource: sagemaker-ssh-helper [arn:aws:sagemaker:eu-west-1:555555555555:notebook-instance/sagemaker-ssh-helper]
8yre1xga6h-algo-1-rmi0c | Initializing new seelog logger
8yre1xga6h-algo-1-rmi0c | New Seelog Logger Creation Complete
8yre1xga6h-algo-1-rmi0c | 2023/04/20 15:05:23 Found config file at /etc/amazon/ssm/amazon-ssm-agent.json.
8yre1xga6h-algo-1-rmi0c | 2023/04/20 15:05:23 processing appconfig overrides
8yre1xga6h-algo-1-rmi0c | Applying config override from /etc/amazon/ssm/amazon-ssm-agent.json.
8yre1xga6h-algo-1-rmi0c | 2023-04-20 15:05:23 WARN Could not read InstanceFingerprint file: InstanceFingerprint does not exist
8yre1xga6h-algo-1-rmi0c | Found config file at /etc/amazon/ssm/amazon-ssm-agent.json.
8yre1xga6h-algo-1-rmi0c | processing appconfig overrides
8yre1xga6h-algo-1-rmi0c | Applying config override from /etc/amazon/ssm/amazon-ssm-agent.json.
8yre1xga6h-algo-1-rmi0c | 2023-04-20 15:05:23 INFO No initial fingerprint detected, generating fingerprint file...
8yre1xga6h-algo-1-rmi0c | 2023-04-20 15:05:23 INFO Successfully registered the instance with AWS SSM using Managed instance-id: mi-1234567890abcdef0
8yre1xga6h-algo-1-rmi0c | Reading package lists...
```

If you see the instance in Fleet Manager, can you check what tags the SSH helper assigned to and and try to connect to the instance with Session Manager from AWS Console?

Also, could you, please, check into what AWS Region AWS cli resolves you local configuration? This command will help you to do that:
```
AWS_PROFILE=default aws configure list | grep region
```

The region should be the same as your notebook instance.

You're correct that `sm-local-ssh-notebook` script is waiting for the SSH Helper to start on remote machine. It expects  instance in SSM with the tag `SSHResourceName` that equials to `<<notebook-instance-name>>`.

Hope it will help you with troubleshooting. Let me know how it goes.

**Comment by mlnrt** [comment]
Hello @ivan-khvostishkov 
Thank you for replying and trying to help and sorry for the late reply. I found time today to try this package and made some progress.
The issue was indeed the region as I has changed my default profile region to do some tests...
That said I am still facing issues following the instructions.
Now the `AWS_PROFILE=default sm-local-ssh-notebook connect <<notebook-instance-name>>` command is working. It is finding the SageMaker notebook registered in SSM as a mi-* instance, saves the key in the .ssh/ folder, etc.
Based on the "Local IDE integration with SageMaker Studio over SSH for PyCharm / VSCode" instructions we are told to follow for the SageMaker notebook case I added the following configuration in the `~/.ssh/config` file:
```
Host sagemkaer-notebook
  HostName localhost
  IdentityFile ~/.ssh/sagemaker-ssh-gw
  Port 10022
  User root
```
It was failing. But then I realized that in the case of the SageMaker notebook, the script configure the forwarding of port 17022, instead of 10022.
Once the port fixed, I can SSH but I am still struggling with VSCode. Looking at the screen capture it seems to establish a remote SSH connection to the SageMaker host, instead of defyning it as a remote SSH interpreter for the commands written in the local notebook. And when connected anyway, I can't see my SageMaker notebook `/home/ec2_user` folder.

The instructions for SageMaker studio also say:
```
You can also configure a remote Jupyter Server as http://127.0.0.1:8889/?token=<<your_token>>.

You will find the full URL with remote token in the [SageMaker_SSH_IDE.ipynb](https://github.com/aws-samples/sagemaker-ssh-helper/blob/main/SageMaker_SSH_IDE.ipynb) notebook in the output after running the cell with sm-ssh-ide start command.
```
Except that for SageMaker notebook I do not see such a URL in the output of the `SageMaker_SSH_Notebook.ipynb` notebook.
And when I do a `jupyter notebook list` in the SageMaker Notebook instance I do not see any Jupyter server running with a token.
Bottom line is I am unable to follow the instructions for a SageMaker notebook case with VSCode.
What am I doing wrong?
Thank you in advance.

**Comment by ivan-khvostishkov** [comment]
Hi, @mlnrt ,

1/ Apologies for the confusion in the documentation. Notebook support is quite limited, as compared to SageMaker Studio. You're correct that you need to use the port 17022, because you cannot use connection to notebook and to SageMaker Studio interchangeably. Unlike `sm-local-ssh-ide`, `sm-local-ssh-notebook` doesn't support VNC or remote notebook connection. Remote paths and directory structures are also different. 

Notebook support is more akin to a training job. It uses [SageMaker local mode](https://sagemaker.readthedocs.io/en/stable/overview.html?#local-mode) to run the container with the SSM agent.

I've added this information to the [FAQ](https://github.com/aws-samples/sagemaker-ssh-helper/blob/main/FAQ.md#are-sagemaker-notebook-instances-supported).

2/ When you say:

> Looking at the screen capture it seems to establish a remote SSH connection to the SageMaker host, instead of defyning it as a remote SSH interpreter for the commands written in the local notebook. 

what do you mean by defining a remote SSH interpreter? Could you, please, point to the VS Code documentation page, so I can be sure we're talking about the same thing?

3/ 

> And when connected anyway, I can't see my SageMaker notebook /home/ec2_user folder.

As mentioned in the [SageMaker_SSH_Notebook.ipynb](https://github.com/aws-samples/sagemaker-ssh-helper/blob/main/SageMaker_SSH_Notebook.ipynb), your notebook instance files will be mounted in the `/opt/ml/input/data/notebook` directory, not in `/home/ec2_user`.

Hope it clarifies your questions.

**Comment by mlnrt** [comment]
Hello @ivan-khvostishkov  sorry it took me so long, I finally found time to try this out again.
To reply about what I mean about defining a remote SSH interpreter is I think what you mean by 

> sm-local-ssh-notebook doesn't support VNC or remote notebook connection

A remote SSH session is opened to the sagemaker notebook instance and everything is ran there. The notebook is not local, sending commands and receiving the results.

I made further tries configuring both VSCode and PyCharm.

- With PyCharm I get the following error:

> Running as root is not recommended. Use --allow-roo to bypass

- With VSCode I can open the remote SSH connection to the sagemaker notebook instance, I can open the notebook which is in `/opt/ml/input/data/notebook` directory, but when I run it, I never see the cells results.

I also don't understand this tip in the instructions:
Tip (PyCharm): Also instead of creating a new venv, point the Python interpreter to the existing location. You can find this location by running a cell with import sys; sys.executable command in a SageMaker Studio notebook. You will get something like /opt/conda/bin/python.
Does this applies to the sagemaker-notebook case?

**Comment by ivan-khvostishkov** [comment]
Hi, @mlnrt , as I understand you have successfully connected with `sm-local-ssh-notebook` to the SageMaker instance container and the port 17022 now is opened on your local machine, allowing you to connect with SSH, right?

It will be very useful if you add screenshots illustrating the remaining problems that you have now, e.g. to illustrate what you mean when you say "I never see the cells results".

If you want to try running Jupyter remotely and connect to it with PyCharm, you will need to pass the `--allow-root` option, as it's done for SageMaker Studio:

https://github.com/aws-samples/sagemaker-ssh-helper/blob/v1.10.4/sagemaker_ssh_helper/sm-ssh-ide#L171 

The tip for determining the Python interpreter is only for SageMaker Studio where you can run the IDE notebook in different images / kernels. The [SageMaker_SSH_Notebook.ipynb](https://github.com/aws-samples/sagemaker-ssh-helper/blob/main/SageMaker_SSH_Notebook.ipynb) runs the container with PyTorch image. As of today, it uses `/opt/conda/bin/python3.8` , so you can use this path.

**Comment by ivan-khvostishkov** [comment]
@mlnrt I will close the issue for now. Please, reopen if you still has this issue.
